### ISSUE TITLE
HashWithIndifferentAccess in the session is converted to Hash during Integration Test

### ISSUE THREAD

**Issue by roberts1000** [rails/rails#26243]
I've run into a bug with an Integration Test.  Hashes with indifferent access aren't maintained between requests.  This problem only happens in testing.  Everything seems to work fine in the live application.  I tried using the bug template, but the template isn't loading the same code as a full Rails app and I'm not sure what the issues is.  (My attempt at the template can be found [here](https://gist.github.com/roberts1000/d5ea11431cc290322cdae70745411ccb)).   I have reproduced this issue in a bare bones 5.0.0.1 app.

To summarize, the integration test does:

````ruby
def test_bug
  get "/one"
  assert_equal(ActiveSupport::HashWithIndifferentAccess, controller.the_hash.class)
  get "/two"
  # This should pass, but it fails because a hash is returned
  assert_equal(ActiveSupport::HashWithIndifferentAccess, controller.the_hash.class)
end
````

The second `assert` fails because a `Hash` instance is returned instead of the original `ActiveSupport::HashWithIndifferentAccess` instance.   Something happens during the call to `get "/two"`; the `HashWithIndifferentAccess` isn't maintained.  The controller actions are dirt simple:

````ruby
def one
  session[:foo] = {}.with_indifferent_access
  render plain: "One"
end
````

````ruby
def two
  render plain: "Two"    
end
````

````ruby
def the_hash
  session[:foo]
end
````
### System configuration
**Rails version**:

I've seen this on 4.2.6, 4.2.7, 5.0.0 and 5.0.0.1.

**Ruby version**:

2.3.1

**Comment by rafaelfranca** [comment]
> This problem only happens in testing.

It shouldn't. If you session is kept in cookies and the cookies are stored as JSON objects (the default since 4.2) the `HashWithIndifferentAccess` instance should not be maintained. In fact, only JSON compatible objects have their types maintained.

Could you reproduce this behavior in a new application?

**Comment by roberts1000** [comment]
Thank you for the feedback.  You are correct.  I doubled checked in a brand new Rails application and the `HashWithIndifferentAccess` is not maintained.  I went back and looked at the application where I initially ran into the issue and we have some code that is causing the `HashWithIndifferentAccess` to be recreated.

I'm going to close the issue since this is definitely not a bug.  Maybe this a potential future feature.
